[clang] Inject IndirectFieldDecl even if name conflicts.

keinflue opened the pull request:
This modifies InjectAnonymousStructOrUnionMembers to inject an IndirectFieldDecl and mark it invalid even if its name conflicts with another name in the scope.

This resolves a crash on a further diagnostic diag::err_multiple_mem_union_initialization which via findDefaultInitializer relies on these declarations being present.

Fixes #149985

keinflue committed:
[clang] Inject IndirectFieldDecl even if name conflicts.

This modifes InjectAnonymousStructOrUnionMembers to inject an IndirectFieldDecl
and mark it invalid even if its name conflicts with another name in the scope.

This resolves a crash on a further diagnostic diag::err_multople_mem_union_initialization
which via findDefaultInitializer relies on these declarations being present.

Fixes #149985

keinflue committed:
Merge branch 'main' into fix-149985

Fznamznon commented on clang/test/CXX/class/class.mem/p13.cpp:
```diff
@@ -63,7 +63,7 @@ struct X4 { // expected-note{{previous}}
     union {
       float Y;
-      unsigned X4; // expected-error{{redeclares 'X4'}}
+      unsigned X4; // expected-error{{redeclares 'X4'}} expected-error {{'X4' has the same name as its class}}
```
I'm not sure it is good to have that type of double erroring,

AaronBallman replied:
Oof. Each are correct, each are helpful in isolation, but having both show up at the same time is a bit chatty. I think the "redeclares" diagnostic is the more informative of the two so I weakly prefer keeping that one and suppressing the other, but I think it's defensible either way.

Fznamznon commented on clang/test/CXX/class/class.union/class.union.anon/p4.cpp:
```diff
@@ -8,3 +8,13 @@ union U {
   };
 };
+
+namespace GH149985 {
```
Can we add a test with reproducer from https://github.com/llvm/llvm-project/issues/149985 ?

AaronBallman replied:
It's situational and a bit of a judgement call; it's good to use the reproducer from the issue because that is easy for reviewers to verify that it actually fixes the bug. But not all reproducers on issues will be reasonable for inclusion in the test suite, so it's fine to make modifications as needed. I think it'd be good to grab the test from the issue in addition to this; the issue is using anonymous structures within a union which means it could exercise somewhat different code paths than your test case.

Fznamznon reviewed (commented):
Adding more reviewers...

erichkeane reviewed (commented):
I think the patch in general is correct but the double-error needs to be dealt with.  I suspect one of them needs to suppress its diagnostic if it is invalid.  While I have a preference for which is first, we should just make sure they both ensure they aren't diagnosing on an invalid IndirectFieldDecl.

keinflue committed:
Implement review suggestions

Disable duplicate diagnostic, include original test case and fix setting only affected fields invalid.

keinflue committed:
Extend release notes

erichkeane commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -5494,55 +5494,62 @@ InjectAnonymousStructOrUnionMembers(Sema &SemaRef, Scope *S, DeclContext *Owner,
                                     RecordDecl *AnonRecord, AccessSpecifier AS,
                                     StorageClass SC,
                                     SmallVectorImpl<NamedDecl *> &Chaining) {
+
```
unrelated change.

erichkeane commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -5494,55 +5494,62 @@ InjectAnonymousStructOrUnionMembers(Sema &SemaRef, Scope *S, DeclContext *Owner,
+      //   distinct from the names of any other entity in the
+      //   scope in which the anonymous union is declared.
+
+      const bool FieldInvalid = CheckAnonMemberRedeclaration(
```
```suggestion
      bool FieldInvalid = CheckAnonMemberRedeclaration(
```

erichkeane commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -5494,55 +5494,62 @@ InjectAnonymousStructOrUnionMembers(Sema &SemaRef, Scope *S, DeclContext *Owner,
+        Chaining.push_back(VD);
+
+      assert(Chaining.size() >= 2);
+      NamedDecl **NamedChain =
```
ooof, no, don't do that!  Use llvm::SmallVector.

keinflue committed:
Implement further stlye-related review suggestions.

keinflue committed:
Merge branch 'main' into fix-149985

keinflue committed:
Fix release notes build

keinflue committed:
Merge branch 'main' into fix-149985
